iroh-rpc/src/stream.rs

```diff
@@ -0,0 +1,227 @@
+use crate::commands::Command;
+use bytecheck::CheckBytes;
+use futures::channel::{mpsc, oneshot};
+use futures::prelude::*;
+use futures::stream::StreamExt;
+use futures::task::{Context, Poll};
+use libp2p::PeerId;
+use log::{debug, error};
+use rkyv;
+use rkyv::{Archive, Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io::BufRead;
+use std::pin::Pin;
+
+use crate::error::RPCError;
+
+/// InStream coordinates a stream of packet data, allowing the user
+/// to receive the chunks of bytes in the correct order.
+// TODO: better name, add timeout, early cancel, and handle errors
+pub struct InStream {
+    packet_receiver: StreamReceiver,
+    out_sender: mpsc::Sender<Command>,
+    chunks: HashMap<u64, Vec<u8>>,
+    id: u64,
+    next_chunk: u64,
+    num_chunks: u64,
+}
+
+impl InStream {
+    pub fn new(
+        header: Header,
+        packet_receiver: StreamReceiver,
+        out_sender: mpsc::Sender<Command>,
+    ) -> Self {
+        InStream {
+            packet_receiver,
+            out_sender,
+            next_chunk: 0,
+            id: header.id,
+            num_chunks: header.num_chunks,
+            chunks: HashMap::new(),
+        }
+    }
+
+    // TODO: this should implement `poll_next` & satisfy the `Stream` trait
```

Comment:
This TODO I want to get to for this PR, the others in this mod, I'd love an opinion on if they can wait an iteration or two.

Reply:
yeah, we will need that for the interface in the other modules

Reply:
wait I already did this, I just never got rid of the original bad implementation. Phew! I thought I was going crazy!